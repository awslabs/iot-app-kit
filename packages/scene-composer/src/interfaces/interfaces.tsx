/**
 * This file contains public interfaces of the scene content the SceneComposer exposes.
 * It is different from the serialization format from the GLTF Toolkit package.
 * These interfaces are used to interact with the SceneComposer instance at
 * runtime, thus the data structure is slightly different from the GLTF Toolkit.
 */

import { ReactElement } from 'react';

import { ISceneNodeInternal } from '../store/Store';

import { ISceneComponent, KnownComponentType } from './components';
import { IRuleBasedMap, ITransform, ITransformConstraint, Vector3 } from './dataTypes';

/************************************************
 * Scene document, entity, component models
 ************************************************/

export interface ISceneDocumentSnapshot {
  /**
   * Serialize the current document snapshot into a document string with the
   * given format version.
   *
   * Note that this interface is not optimized for very large scene, which
   * we may introduce new APIs later to use stream API to handle the output.
   *
   * TODO: do we want to allow updating the document version?
   * [Option 1]: Externalize the document version instea of keeping the doucment version in the document itself.
   * [Option 2]: Allow passing in a parameter to set the document version.
   * [Option 3]: Automatically increment the document version by 1 based on the original input document version.
   *
   * We'll go with option 1 for now for simplicity.
   *
   * TODO: what should we return when the scene is empty?
   * [Option 1]: Return a falsy value, such as empty string.
   * [Option 2]: Return a valid scene with no entity.
   *
   * We'll go with option 2 for now.
   *
   * @param specVersion - version of the document format
   */
  serialize(specVersion: string, ...stringifyArgs): string;
}

export interface ISceneDocument {
  nodeMap: Record<string, ISceneNode>;
  ruleMap: Record<string, IRuleBasedMap>;
  rootNodeRefs: string[];
  unit?: string;
  version: string;
  properties?: Partial<Record<KnownSceneProperty, any>>;
  specVersion?: string;
}

export interface ISceneNode {
  // ref is optional when used in APIs to create nodes
  ref?: string;
  name?: string;
  transform?: ITransform;
  transformConstraint?: ITransformConstraint;
  components?: ISceneComponent[];
  parentRef?: string;
  childRefs?: string[];
  properties?: Partial<Record<'alwaysVisible' | 'matterportId' | 'hiddenWhileImmersive', boolean | string>>;
}

export enum KnownSceneProperty {
  BaseUrl = 'baseUrl',
  EnvironmentPreset = 'environmentPreset',
  DataBindingConfig = 'dataBindingConfig',
  ComponentSettings = 'componentSettings',
  MatterportModelId = 'matterportModelId',
  LayerIds = 'layerIds',
  SceneRootEntityId = 'sceneRootEntityId',
  TagCustomColors = 'tagCustomColors',
  FogSettings = 'fogSettings',
}

/************************************************
 * Scene Composer Control APIs
 ************************************************/

export type CameraControlMode = 'transition' | 'teleport';
export type CameraControlsType = 'orbit' | 'pan' | 'immersive' | 'pointerLock';
export type TransformControlMode = 'translate' | 'rotate' | 'scale';
export enum CameraViewAxisValues {
  Front = 'front',
  Back = 'back',
  Left = 'left',
  Right = 'right',
  Top = 'top',
  Bottom = 'bottom',
}

// an object ref or a precise camera setting
export type FixedCameraTarget = { position: Vector3; target: Vector3 };
export type NamedCameraTarget = string;
export type CameraTarget = NamedCameraTarget | FixedCameraTarget;

export type URIModifier = (uri: string) => string;

export type CreateErrorViewCallback = (error: Error) => ReactElement;

/************************************************
 * Scene Namespace
 ************************************************/

const IotTwinMakerNamespace = 'iottwinmaker.common';
export const IotTwinMakerNamespaceSeparator = ':';
export const IotTwinMakerIconNamespace = `${IotTwinMakerNamespace}.icon`;
export const IotTwinMakerColorNamespace = `${IotTwinMakerNamespace}.color`;
export const IotTwinMakerNumberNamespace = `${IotTwinMakerNamespace}.number`;
export const IotTwinMakerOpacityNamespace = `${IotTwinMakerNamespace}.opacity`;
/************************************************
 * Scene Resource
 ************************************************/

export enum SceneResourceType {
  Icon = 'Icon',
  Color = 'Color',
  Number = 'Number',
  Opacity = 'Opacity',
}

export interface SceneResourceInfo {
  type: SceneResourceType;
  value: string;
}

export interface AddingWidgetInfo {
  type?: KnownComponentType;
  node: ISceneNodeInternal;
}

export interface MeshStyle {
  color?: string | number;
  opacity?: number;
  transparent?: boolean;
}

export interface StyleTarget {
  dataBindingContext: unknown;
  style: MeshStyle;
}

export interface IFogSettings {
  colorHex: number;
  near: number;
  far: number;
}
