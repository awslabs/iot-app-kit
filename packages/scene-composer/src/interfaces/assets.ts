export enum AssetType {
  BIN = 'BIN',
  GLB = 'GLB',
  GLTF = 'GLTF',
  JPG = 'JPG',
  JPEG = 'JPEG',
  MP3 = 'MP3',
  MP4 = 'MP4',
  PDF = 'PDF',
  PNG = 'PNG',
  TILES_3D = 'Tiles3D',
}

export const ModelFileTypeList: AssetType[] = [AssetType.GLB, AssetType.GLTF, AssetType.TILES_3D];
export const TextureFileTypeList: AssetType[] = [AssetType.JPG, AssetType.JPEG, AssetType.PNG];
