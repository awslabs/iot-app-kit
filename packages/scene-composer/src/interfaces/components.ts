import * as SceneModels from '../models/SceneModels';

/************************************************
 * Components
 ************************************************/

import { DistanceUnit, INavLink, IValueDataBinding, Vector3 } from './dataTypes';

export enum KnownComponentType {
  ModelRef = 'ModelRef',
  SubModelRef = 'SubModelRef',
  Camera = 'Camera',
  Light = 'Light',

  Tag = 'Tag',
  ModelShader = 'ModelShader',
  MotionIndicator = 'MotionIndicator',
  DataOverlay = 'DataOverlay',
}

export interface ISceneComponent {
  // ref is optional when used in APIs to create components
  ref?: string;
  type: KnownComponentType | string;
}

export interface IModelRefComponent extends ISceneComponent {
  uri: string;
  modelType: string;

  unitOfMeasure?: DistanceUnit;
  localScale?: Vector3;
  castShadow?: boolean;
  receiveShadow?: boolean;
}

export interface ISubModelRefComponent extends ISceneComponent {
  parentRef?: string;
  selector: string | number;
}

export interface IAnchorComponent extends ISceneComponent {
  icon?: string;
  valueDataBinding?: IValueDataBinding;
  ruleBasedMapId?: string;
  navLink?: INavLink;
  offset?: Vector3;
}

/**
 * Default Anchor status strings
 */
export enum DefaultAnchorStatus {
  Info = 'Info',
  Warning = 'Warning',
  Error = 'Error',
  Video = 'Video',
}

/**
 * Selected Anchor string
 */
export const SelectedAnchor = 'Selected';

/**
 * Additional Tag Component Data to be given when a tag node is clicked or changed
 */
export interface ITagData {
  navLink?: INavLink;
  dataBindingContext?: unknown;
}

/**
 * Type that can be represented by different additional component data types such as ITagData | IFutureComponentData
 */
export type AdditionalComponentData = ITagData;

/**
 * Callback signature for selection of with Widgets.
 */
export interface ISelectionChangedEvent {
  componentTypes: KnownComponentType[] | string[];
  nodeRef?: string;
  additionalComponentData?: AdditionalComponentData[];
}

/**
 * Callback signature for interaction with Widgets.
 */
export interface IWidgetClickEvent {
  componentTypes: KnownComponentType[] | string[];
  nodeRef: string;
  additionalComponentData?: AdditionalComponentData[];
}

export type WidgetClickEventCallback = (event: IWidgetClickEvent) => void;

export type SelectionChangedEventCallback = (event: ISelectionChangedEvent) => void;

export interface ILightComponent extends ISceneComponent, SceneModels.Component.Light {}

export type CameraType = 'Perspective' | 'Orthographic';

export interface ICameraBasics {
  cameraType: CameraType;
  fov?: number;
  near: number;
  far: number;
  zoom: number;
}

export interface CameraSettings extends ICameraBasics {
  transform: SceneModels.Transform;
}

export interface ICameraComponent extends ISceneComponent, ICameraBasics {}

export interface IColorOverlayComponent extends ISceneComponent {
  valueDataBinding?: IValueDataBinding;
  ruleBasedMapId?: string;
}

export interface IMotionIndicatorComponent extends ISceneComponent, SceneModels.Component.MotionIndicator {}

export interface IDataOverlayComponent extends ISceneComponent, SceneModels.Component.DataOverlay {}
