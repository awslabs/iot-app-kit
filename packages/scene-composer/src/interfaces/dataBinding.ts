/************************************************
 * Data Input
 ************************************************/

// Timestamp should be millisecond epoch time
export type Timestamp = number;

// Inherit the Grafana convention of this "valueType" for time
export const DataFieldTimeType = 'time';

export interface IDataInput {
  dataFrames: IDataFrame[];

  // The current time to show, if it's undefined, the last data points will be
  // used
  timeCursor?: Timestamp;

  // The total range of the data
  // NOTE: Grafana treat both from and to inclusive, but we don't define
  // the behavior explicitly here due to the fact that some data source
  // may not support it natively. It should not matter in most of the use
  // cases we handle.
  timeRange: {
    from: Timestamp;
    to: Timestamp;
  };
}

export interface IDataFrame {
  // This maps to the refId in grafana
  dataFrameId: string;
  // This maps to the fields in grafana, fields are like columns in table
  // fields are assumed to have the same length in a single data frame.
  // For time series data, the first field should be Time. (by convention)
  fields: IDataField[];
}

export interface IDataField {
  // This is the column name, by convention it should be mapped to property name
  name: string;

  // labels are used to identify an instance of a multi-dimentional query,
  // e.g. query all the instances of the servers and return each server's data
  // as a separate field, labels can be used to uniquely identify each
  // individual server instances.
  // see: https://grafana.com/docs/grafana/latest/developers/plugins/data-frames/
  labels?: Record<string, string>;

  // Data type of the values
  valueType: ValueType;

  // The actual data records
  values: any[];
}

// We'll support the following for now. We can extend the list in the future
// 'time' data should be the same format as Timestamp data type
export type ValueType = 'number' | 'string' | 'boolean' | 'time';

export type {
  IDataFieldOption,
  IDataBindingConfig,
  IDataBindingTemplate,
  IValueDataBindingStore,
  IValueDataBindingProvider,
  IValueDataBindingProviderState,
} from '@iot-app-kit/source-iottwinmaker';

export interface ISelectedDataBinding {
  entityId: string;
  componentName?: string;
}
