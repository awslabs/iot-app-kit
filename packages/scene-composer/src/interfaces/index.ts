export * from './assets';
export * from './interfaces';
export * from './components';
export * from './dataTypes';
export * from './dataBinding';
export * from './metricRecorder';
export * from './feature';
export * from './sceneViewer';
export * from './componentSettings';
