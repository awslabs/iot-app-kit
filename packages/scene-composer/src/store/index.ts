export * from './StoreOperations';
export * from './Store';
