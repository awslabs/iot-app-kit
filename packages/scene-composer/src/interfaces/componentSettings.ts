import { KnownComponentType } from './components';

export interface ITagSettings {
  scale: number;
  autoRescale: boolean;
}

export interface IOverlaySettings {
  overlayPanelVisible?: boolean;
}

export type IComponentSettings = ITagSettings | IOverlaySettings;
export type IComponentSettingsMap = Record<KnownComponentType | string, IComponentSettings>;
