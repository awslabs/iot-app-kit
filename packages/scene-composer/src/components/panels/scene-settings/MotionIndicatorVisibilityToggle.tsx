import React, { useContext, useMemo } from 'react';
import { useIntl } from 'react-intl';
import { Box, Toggle } from '@awsui/components-react';

import { useStore, useViewOptionState } from '../../../store';
import { sceneComposerIdContext } from '../../../common/sceneComposerIdContext';
import { KnownComponentType } from '../../../interfaces';

export const MotionIndicatorVisibilityToggle: React.FC = () => {
  const sceneComposerId = useContext(sceneComposerIdContext);
  const { motionIndicatorVisible, toggleMotionIndicatorVisibility } = useViewOptionState(sceneComposerId);
  const getComponentRefByType = useStore(sceneComposerId)((state) => state.getComponentRefByType);
  const componentNodeMap = useStore(sceneComposerId)((state) => state.document.componentNodeMap);
  const { formatMessage } = useIntl();

  const hasMotionIndicator = useMemo(() => {
    return Object.keys(getComponentRefByType(KnownComponentType.MotionIndicator)).length > 0;
  }, [componentNodeMap, getComponentRefByType]);

  return (
    <React.Fragment>
      <Box variant='p' fontWeight='bold' margin={{ bottom: 'xxs' }}>
        {formatMessage({ description: 'Sub section label', defaultMessage: 'Motion indicator' })}
      </Box>
      <Toggle
        disabled={!hasMotionIndicator}
        checked={motionIndicatorVisible}
        onChange={toggleMotionIndicatorVisibility}
      >
        {formatMessage({ description: 'Toggle label', defaultMessage: 'Visibility' })}
      </Toggle>
    </React.Fragment>
  );
};
