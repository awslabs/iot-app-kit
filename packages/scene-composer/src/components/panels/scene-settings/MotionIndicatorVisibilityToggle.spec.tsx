import { render } from '@testing-library/react';
import React from 'react';

import { KnownComponentType } from '../../../interfaces';
import { useStore } from '../../../store';

import { MotionIndicatorVisibilityToggle } from './MotionIndicatorVisibilityToggle';

describe('MotionIndicatorVisibilityToggle', () => {
  const getComponentRefByType = jest.fn();

  const createState = (visible: boolean) => ({
    noHistoryStates: {
      ...useStore('default').getState().noHistoryStates,
      motionIndicatorVisible: visible,
      toggleMotionIndicatorVisibility: jest.fn(),
    },
    getComponentRefByType,
  });

  it('should render correctly', async () => {
    getComponentRefByType.mockReturnValue({ type: KnownComponentType.MotionIndicator });
    useStore('default').setState(createState(true));
    const { container } = render(<MotionIndicatorVisibilityToggle />);

    expect(container).toMatchSnapshot();
  });
});
