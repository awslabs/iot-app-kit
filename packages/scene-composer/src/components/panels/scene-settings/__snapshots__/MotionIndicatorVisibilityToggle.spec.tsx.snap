// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MotionIndicatorVisibilityToggle should render correctly 1`] = `
<div>
  <div
    data-mocked="Box"
    font-weight="bold"
    margin="{\\"bottom\\":\\"xxs\\"}"
    variant="p"
  >
    Motion indicator
  </div>
  <div
    data-mocked="Toggle"
  >
    Visibility
  </div>
</div>
`;
